doc/design/python_interface.md

```diff
@@ -0,0 +1,101 @@
+# Python Interface Design
+
+## Variable as variable data type (not parameter)
+The class Variable is introduced to define the variables that are not parameters.
+
+All the variables are defined in a variable scope, `VarScope`, it will check duplicate variable names.
+
+```python
+class VarScope(object):
+    '''
+    VarScope is names' container of VarDesc in compile period.
+    '''
+    pass
+```
+
+The `Variable`'s definition is as follows
+```python
+class _unique_name_generator:
+    counter = 0
+
+    def __call__(self, prefix):
+        name = '%s-%d' % (prefix, counter)
```

Comment:
Maybe `%s.%d` is better. Because `%s-%d` is not a valid filename on some platforms, e.g., Windows.